receiver/factory.go

```diff
@@ -71,6 +59,37 @@ type Factory interface {
 //   An empty interface wrapping a pointer to the config struct to unmarshal into.
 type CustomUnmarshaler func(v *viper.Viper, viperKey string, sourceViperSection *viper.Viper, intoCfg interface{}) error
 
+// Factory can create TraceReceiver and MetricsReceiver.
+type Factory interface {
+	BaseFactory
+
+	// CreateTraceReceiver creates a trace receiver based on this config.
+	// If the receiver type does not support tracing or if the config is not valid
+	// error will be returned instead.
+	CreateTraceReceiver(ctx context.Context, logger *zap.Logger, cfg configmodels.Receiver,
```

Comment:
not for this PR but something to consider down the road: perhaps instead of having separate Create methods for metrics/trace it could be a single one and later when connecting it to pipes the configuration checks if it supports the proper consumer interface. This removes the need for the tracking that receivers that support both metrics and traces need to do and the dead implementation of Create for the data types not supported by the receiver.

Reply:
Side note we may have a bug for receivers that support both data types if only one has an actual pipeline, will check it later.

Reply:
I am not sure combining receivers is easily possible. The receivers accept a consumer when they are created. When creating the receiver the appropriate consumer type has to be provided. Unless we also merge the consumer interface we will have to keep the "Create" functions of factories separate per data type. There are probably other ways, I agree it is worth exploring if further simplification and unification of interfaces is possible.

As for dual-datatype receivers, we have OpenCensus and it is often used with a single pipeline. Ping me if you find a bug in this setting.

Reply:
Yeah, good point about the consumer type. Regarding the possible bug, it depends on the implementation of the receiver OC should not have it.

Reply:
OK, resolving this for now.